tests/frame_manager_test.go

```diff
@@ -33,37 +35,79 @@ func TestWaitForFrameNavigationWithinDocument(t *testing.T) {
 	for _, tc := range testCases {
 		tc := tc
 		t.Run(tc.name, func(t *testing.T) {
-			tb := newTestBrowser(t, withFileServer())
-			p := tb.NewPage(nil)
+			t.Parallel()
 
-			resp := p.Goto(tb.staticURL("/nav_in_doc.html"), nil)
-			require.NotNil(t, resp)
+			errc := make(chan error, 1)
+			go func() {
+				tb := newTestBrowser(t, withFileServer())
+				p := tb.NewPage(nil)
 
-			// A click right away could possibly trigger navigation before we
-			// had a chance to call WaitForNavigation below, so give it some
-			// time to simulate the JS overhead, waiting for XHR response, etc.
-			time.AfterFunc(timeout*time.Millisecond, func() { //nolint:durationcheck
-				_ = tb.await(func() error {
-					_ = p.Click(tc.selector, nil)
-					return nil
-				})
-			})
+				resp := p.Goto(tb.staticURL("/nav_in_doc.html"), nil)
+				require.NotNil(t, resp)
 
-			done := make(chan struct{}, 1)
-			go func() {
-				require.NotPanics(t, func() {
-					p.WaitForNavigation(tb.toGojaValue(&common.FrameWaitForNavigationOptions{
-						Timeout: timeout * 3, // interpreted as ms
+				// A click right away could possibly trigger navigation before we
+				// had a chance to call WaitForNavigation below, so give it some
+				// time to simulate the JS overhead, waiting for XHR
+				// response, etc.
+				<-time.After(timeout * time.Millisecond) //nolint:durationcheck
+
+				// if one of the promises panics, err will contain the error
+				errc <- tb.await(func() error {
+					_ = p.Click(tc.selector, nil)
+					_ = p.WaitForNavigation(tb.toGojaValue(&common.FrameWaitForNavigationOptions{
+						Timeout: 3 * timeout, // interpreted as ms
 					}))
+					return nil
 				})
-				done <- struct{}{}
 			}()
-
 			select {
-			case <-done:
-			case <-time.After(timeout * 5 * time.Millisecond): //nolint:durationcheck
+			case err := <-errc:
+				require.NoError(t, err)
+			case <-time.After(5 * timeout * time.Millisecond): //nolint:durationcheck
 				t.Fatal("Test timed out")
 			}
 		})
 	}
 }
+
+func TestWaitForFrameNavigation(t *testing.T) {
+	tb := newTestBrowser(t, withHTTPServer())
+	p := tb.NewPage(nil)
+
+	tb.withHandler("/first", func(w http.ResponseWriter, _ *http.Request) {
+		fmt.Fprintf(w, `
+		<html>
+			<head>
+				<title>First page</title>
+			</head>
+			<body>
+				<a href="/second">click me</a>
+			</body>
+		</html>
+		`)
+	})
+	tb.withHandler("/second", func(w http.ResponseWriter, _ *http.Request) {
+		fmt.Fprintf(w, `
+		<html>
+			<head>
+				<title>Second page</title>
+			</head>
+			<body>
+				<a href="/first">click me</a>
+			</body>
+		</html>
+		`)
+	})
+
+	require.NotNil(t, p.Goto(tb.URL("/first"), tb.toGojaValue(&common.FrameGotoOptions{
+		WaitUntil: common.LifecycleEventNetworkIdle,
+		Timeout:   common.DefaultTimeout,
+	})))
+	err := tb.await(func() error {
+		_ = p.Click(`a`, nil)
+		p.WaitForNavigation(nil)
+		return nil
+	})
+	require.NoError(t, err)
+	require.Equal(t, p.Title(), "Second page")
```

Comment:
Nit: the second argument to `Equal()` (and most of the `testify` methods) is the expected value. Otherwise it will result in confusing Actual/Expected errors.

Reply:
Resolved in: 262eaeed1765e1599b6bdf82d1f2ceabd6eb5cb3